text/3697-declarative-attribute-macros.md

```diff
@@ -0,0 +1,155 @@
+the user to enable a feature.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When defining a `macro_rules!` macro, you can prefix some of the macro's rules
+with `attr(...)` to allow using the macro as an attribute. The
+arguments to the attribute, if any, are parsed by the *MacroMatcher* in the
+first set of parentheses; the second *MacroMatcher* parses the entire construct
+the attribute was applied to. The resulting macro will work anywhere an
+attribute currently works.
+
+```rust
+macro_rules! main {
+    attr() ($func:item) => { make_async_main!($func) };
+    attr(threads = $threads:literal) ($func:item) => { make_async_main!($threads, $func) };
+}
+
+#[main]
+async fn main() { ... }
+
+#[main(threads = 42)]
+async fn main() { ... }
+```
+
+Attribute macros defined using `macro_rules!` follow the same scoping rules as
+any other macro, and may be invoked by any path that resolves to them.
+
+An attribute macro must not require itself for resolution, either directly or
+indirectly (e.g. applied to a containing module or item).
+
+Note that a single macro can have both `attr` and non-`attr` rules. Attribute
+invocations can only match the `attr` rules, and non-attribute invocations can
+only match the non-`attr` rules. This allows adding `attr` rules to an existing
+macro without breaking backwards compatibility.
+
+For simplicity, an attribute macro may not recursively invoke its `attr` rules;
+to recurse, invoke a non-`attr` rule or another macro.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The grammar for macros is extended as follows:
+
+> _MacroRule_ :\
+> &nbsp;&nbsp; ( `attr` _MacroMatcher_ )<sup>?</sup>  _MacroMatcher_ `=>` _MacroTranscriber_
+
+The first _MacroMatcher_ matches the attribute's arguments, which will be an
+empty token tree if either not present (`#[myattr]`) or empty (`#[myattr()]`).
+The second _MacroMatcher_ matches the entire construct the attribute was
+applied to, receiving precisely what a proc-macro-based attribute would in the
+same place.
+
+Only a rule matching both the arguments to the attribute and the construct the
+attribute was applied to will apply. Note that the captures in both
+`MacroMatcher`s share the same namespace; attempting to use the same name for
+two captures will give a "duplicate matcher binding" error.
+
+This grammar addition is backwards compatible: previously, a _MacroRule_ could
+only start with `(`, `[`, or `{`, so the parser can easily distinguish the
+identifier `attr`.
+
+Attribute macros declared using `macro_rules!` are
+[active](https://doc.rust-lang.org/reference/attributes.html#active-and-inert-attributes),
+just like those declared using proc macros.
+
+Adding `attr` rules to an existing macro is a semver-compatible change.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
```

Comment:
Looking over https://veykril.github.io/tlborm/decl-macros/building-blocks/parsing.html#function, I see people quickly being frustrated with this because of how hard attribute macros for functions are to write or frustrated with dealing with buggy attribute macros for functions.

Besides the complexity, there are the footnotes
- What was left out
- That you can't correctly parse the full argument syntax supported by Rust.

On one hand, this makes me worry that `macro_rules!` isn't mature enough for this feature and we should wait.  On the other hand, this might be the motivation for improving it.

I think this is less the case with #3698
- those aren't geared towards functions
- structs are relatively easy
- enums are easier than that post has to deal with because derives append, rather than replace

Reply:
I agree. One day I tried to parse an `impl` block to add a bunch of attributes to every method in that block, but I quickly gave up because I just couldn't parse an `impl` block with generics. Example:

```rust
macro_rules! parse_impl {
    (
        impl $(< $($generics:tt)* >)? $self_ty:ty {
            $($item:item)*
        }
    ) => {};
}

parse_impl! {
    impl<T> Foo<T> {}
}
```
This fails with a cryptic error message:
```
local ambiguity when calling macro `parse_impl`: multiple parsing options: built-in NTs ty ('self_ty') or 1 other option
```
(I don't even know what NT means 😳😳).

The [real underlying use case](https://elastio.github.io/bon/guide/patterns/shared-configuration#associated-methods) for this was researching a way for the users my proc-macro library to share common configurations between different macro invocations using the `macro_rules_attribute` crate.

Reply:
As with https://github.com/rust-lang/rfcs/pull/3698#discussion_r1779326889, I think the RFC needs to be looked at with the assumption that we will separately be trying to improve parsing. Maybe it should be listed as a future possibility just to note that it's under consideration?

> (I don't even understand what NT means 😳😳).

Nonterminal, but yes that is my least favorite macro messsage :)

Reply:
I actually expect attributes to be easier in more cases than derives. Derives will more often need to look inside the struct/enum and deal with the full variety of ADT declaration syntax. Meanwhile, many attributes seem more likely to only work on specific things (e.g. functions with a specific signature) and often ignore the body entirely.